### ISSUE TITLE
terrible performance of `fib` function

### ISSUE THREAD

**Issue by aaaaargZombies** [aaaaargZombies/craptography#1]
## Description

I thought it would be quite slow with a naive recursive implementation of a fibonacci sequence, but I was in for a bit of a shock when I tried to run something more than a couple of words.

https://github.com/aaaaargZombies/craptography/blob/2ed50cb69a66065defcf6750cbf82263d9e777c9/src/Crapto.hs#L80-L97

Test text used:

```
apl
Lorem ipsum dolor sit amet, qui minim labore 
```

Abridged profile output

```
	Tue Apr 30 18:05 2024 Time and Allocation Profiling Report  (Final)

	   craptography-exe +RTS -N -p -RTS -i test.txt -o out.txt

	total time  =       48.45 secs   (165216 ticks @ 1000 us, 12 processors)
	total alloc =     299,240 bytes  (excludes profiling overheads)

COST CENTRE MODULE           SRC                           %time %alloc

fib.fp      Crapto           src/Crapto.hs:(89,3)-(92,41)  100.0    0.0
MAIN        MAIN             <built-in>                      0.0   60.2
main        Main             app/Main.hs:(64,1)-(78,9)       0.0   19.8
rotFib      Crapto           src/Crapto.hs:(67,1)-(71,18)    0.0    2.8
CAF         GHC.IO.Handle.FD <entire-module>                 0.0   11.6


                                                                                                      individual      inherited
COST CENTRE                   MODULE                  SRC                          no.     entries  %time %alloc   %time %alloc

MAIN                          MAIN                    <built-in>                   238           0    0.0   60.2   100.0  100.0
 main                         Main                    app/Main.hs:(64,1)-(78,9)    477           0    0.0   19.8   100.0   24.4
  help                        Main                    app/Main.hs:13:5-8           478           1    0.0    0.0     0.0    0.0
  inFile                      Main                    app/Main.hs:10:5-10          483           1    0.0    0.0     0.0    0.0
  main.opts                   Main                    app/Main.hs:66:7-44          479           1    0.0    0.0     0.0    0.1
   parseArguments             Main                    app/Main.hs:(22,1)-(48,15)   480           5    0.0    0.0     0.0    0.1
    parseArguments.updateOpts Main                    app/Main.hs:(26,3)-(48,15)   481           4    0.0    0.1     0.0    0.1
  main.rotate                 Main                    app/Main.hs:67:7-58          485           1    0.0    0.0   100.0    4.6
   decrapt                    Main                    app/Main.hs:12:5-11          486           1    0.0    0.0     0.0    0.0
   rotFib                     Crapto                  src/Crapto.hs:(67,1)-(71,18) 487           1    0.0    2.8   100.0    4.6
    rotFib.\                  Crapto                  src/Crapto.hs:70:31-50       488          50    0.0    0.6   100.0    1.8
     rotFibChar               Crapto                  src/Crapto.hs:58:1-30        489          50    0.0    0.9   100.0    1.2
      rotChar                 Crapto                  src/Crapto.hs:(51,1)-(55,18) 490          50    0.0    0.0     0.0    0.0
       isLower                Crapto                  src/Crapto.hs:17:1-40        491          50    0.0    0.0     0.0    0.0
       isUpper                Crapto                  src/Crapto.hs:14:1-40        501          12    0.0    0.0     0.0    0.0
       isDigit                Crapto                  src/Crapto.hs:20:1-25        504           0    0.0    0.0     0.0    0.0
       lowerRot               Crapto                  src/Crapto.hs:48:1-36        494           0    0.0    0.0     0.0    0.0
        alphabetRot           Crapto                  src/Crapto.hs:39:1-89        499          38    0.0    0.0     0.0    0.0
         indexOf              Crapto                  src/Crapto.hs:(35,1)-(36,59) 500         467    0.0    0.0     0.0    0.0
       upperRot               Crapto                  src/Crapto.hs:45:1-36        507           0    0.0    0.0     0.0    0.0
        alphabetRot           Crapto                  src/Crapto.hs:39:1-89        508           1    0.0    0.0     0.0    0.0
         indexOf              Crapto                  src/Crapto.hs:(35,1)-(36,59) 509          12    0.0    0.0     0.0    0.0
      fib                     Crapto                  src/Crapto.hs:(81,1)-(97,41) 496           0    0.0    0.0   100.0    0.2
       fib.f                  Crapto                  src/Crapto.hs:(83,3)-(87,16) 497          39    0.0    0.2   100.0    0.2
        fib.fp                Crapto                  src/Crapto.hs:(89,3)-(92,41) 498 24599519259  100.0    0.0   100.0    0.0
  outFile                     Main                    app/Main.hs:11:5-11          484           1    0.0    0.0     0.0    0.0
```

## Acceptance Criteria

- [ ] I shouldn't be wondering how long it's going to take. <500ms for 1000 lines of text would still seem slow based on feel.

## Implementation notes

I can think of two options off the top of my head

1. momoization, which I'm not sure how to do in Haskell with immutable data  (but will be interesting to investigate)
2. a scan where I pass the last and current `fib` val to the next function to compute the next in the sequence rather than recomputing the bottom layers. Should be easy in Haskell but means I can't start the function off at any point in the sequence. This might be an issue when it comes to `stdin` where I imagine I'll work on a line at a time as it arrives, it also means I won't be able to split up a large file into parallel chunks. Though these are both features I haven't added yet and may not do.

**Comment by aaaaargZombies** [comment]
sort of truncated scan / fold thing

```haskell
rotFib :: String -> String
rotFib msg = go (left, right, [rotChar 1 h]) t
 where
  left = 0
  right = 1
  (h : t) = msg

  go :: (Int, Int, String) -> String -> String
  go (_, _, acc) [] = acc
  go (l, r, acc) (x : xs) =
    let n = (l + r)
        a = rotChar n x
     in go (r, n, acc <> [a]) xs
```

[view code in context](https://github.com/aaaaargZombies/craptography/blob/1a33b9b0b46e22359f765eb1d08f54449118a948/src/Crapto.hs#L57-L69)

very large improvement but still not as fast as I'd hoped

---

re-run of the first test

```
	total time  =        0.00 secs   (14 ticks @ 1000 us, 12 processors)
	total alloc =     354,416 bytes  (excludes profiling overheads)
```


1000 lines / 80 col lorem ipsum

```
	Wed May  1 13:42 2024 Time and Allocation Profiling Report  (Final)

	   craptography-exe +RTS -N -p -RTS -i test.md -o out.txt

	total time  =       22.53 secs   (76841 ticks @ 1000 us, 12 processors)
	total alloc = 120,035,902,544 bytes  (excludes profiling overheads)

COST CENTRE MODULE SRC                           %time %alloc

rotFib.go   Crapto src/Crapto.hs:(65,3)-(69,32)   99.7  100.0


                                                                                                      individual      inherited
COST CENTRE                   MODULE                  SRC                          no.     entries  %time %alloc   %time %alloc

MAIN                          MAIN                    <built-in>                   210           0    0.0    0.0   100.0  100.0
 CAF                          GHC.Conc.Signal         <entire-module>              395           0    0.0    0.0     0.0    0.0
 CAF                          GHC.Event.Thread        <entire-module>              380           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Encoding         <entire-module>              361           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Encoding.Iconv   <entire-module>              359           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Exception        <entire-module>              353           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.FD               <entire-module>              352           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Handle.FD        <entire-module>              350           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Handle.Internals <entire-module>              349           0    0.0    0.0     0.0    0.0
 CAF:defaultOpts              Main                    app/Main.hs:18:1-11          275           0    0.0    0.0     0.0    0.0
  defaultOpts                 Main                    app/Main.hs:18:1-86          426           1    0.0    0.0     0.0    0.0
 CAF:digits                   Crapto                  src/Crapto.hs:32:1-6         220           0    0.0    0.0     0.0    0.0
  digits                      Crapto                  src/Crapto.hs:32:1-21        458           1    0.0    0.0     0.0    0.0
 CAF:isDigit                  Crapto                  src/Crapto.hs:20:1-7         225           0    0.0    0.0     0.0    0.0
  isDigit                     Crapto                  src/Crapto.hs:20:1-25        456           1    0.0    0.0     0.0    0.0
 CAF:lowerAlphabet            Crapto                  src/Crapto.hs:26:1-13        222           0    0.0    0.0     0.0    0.0
  lowerAlphabet               Crapto                  src/Crapto.hs:26:1-28        437           1    0.0    0.0     0.0    0.0
 CAF:lowerRot                 Crapto                  src/Crapto.hs:48:1-8         234           0    0.0    0.0     0.0    0.0
  lowerRot                    Crapto                  src/Crapto.hs:48:1-36        448           1    0.0    0.0     0.0    0.0
 CAF:lvl11_r2HF               Main                    <no location info>           281           0    0.0    0.0     0.0    0.0
 CAF:lvl13_r2HH               Main                    <no location info>           282           0    0.0    0.0     0.0    0.0
 CAF:lvl15_r2HJ               Main                    <no location info>           283           0    0.0    0.0     0.0    0.0
 CAF:lvl1_r2Hv                Main                    <no location info>           276           0    0.0    0.0     0.0    0.0
 CAF:lvl3_r2Hx                Main                    <no location info>           277           0    0.0    0.0     0.0    0.0
 CAF:lvl5_r2Hz                Main                    <no location info>           278           0    0.0    0.0     0.0    0.0
 CAF:lvl7_r2HB                Main                    <no location info>           279           0    0.0    0.0     0.0    0.0
 CAF:lvl9_r2HD                Main                    <no location info>           280           0    0.0    0.0     0.0    0.0
 CAF:main1                    Main                    <no location info>           298           0    0.0    0.0     0.0    0.0
  main                        Main                    app/Main.hs:(64,1)-(78,9)    420           1    0.0    0.0     0.0    0.0
 CAF:upperAlphabet            Crapto                  src/Crapto.hs:29:1-13        221           0    0.0    0.0     0.0    0.0
  upperAlphabet               Crapto                  src/Crapto.hs:29:1-28        440           1    0.0    0.0     0.0    0.0
 CAF:upperRot                 Crapto                  src/Crapto.hs:45:1-8         233           0    0.0    0.0     0.0    0.0
  upperRot                    Crapto                  src/Crapto.hs:45:1-36        441           1    0.0    0.0     0.0    0.0
 main                         Main                    app/Main.hs:(64,1)-(78,9)    421           0    0.0    0.0   100.0  100.0
  help                        Main                    app/Main.hs:13:5-8           422           1    0.0    0.0     0.0    0.0
  inFile                      Main                    app/Main.hs:10:5-10          427           1    0.0    0.0     0.0    0.0
  main.opts                   Main                    app/Main.hs:66:7-44          423           1    0.0    0.0     0.0    0.0
   parseArguments             Main                    app/Main.hs:(22,1)-(48,15)   424           5    0.0    0.0     0.0    0.0
    parseArguments.updateOpts Main                    app/Main.hs:(26,3)-(48,15)   425           4    0.0    0.0     0.0    0.0
  main.rotate                 Main                    app/Main.hs:67:7-58          429           1    0.0    0.0   100.0  100.0
   decrapt                    Main                    app/Main.hs:12:5-11          430           1    0.0    0.0     0.0    0.0
   rotFib                     Crapto                  src/Crapto.hs:(58,1)-(69,32) 431           1    0.0    0.0   100.0  100.0
    rotFib.go                 Crapto                  src/Crapto.hs:(65,3)-(69,32) 434       65473   99.7  100.0   100.0  100.0
     rotFib.go.a              Crapto                  src/Crapto.hs:68:9-23        445       65472    0.0    0.0     0.2    0.0
      rotChar                 Crapto                  src/Crapto.hs:(51,1)-(55,18) 446       65472    0.1    0.0     0.2    0.0
       isLower                Crapto                  src/Crapto.hs:17:1-40        447       65472    0.1    0.0     0.1    0.0
       isUpper                Crapto                  src/Crapto.hs:14:1-40        455       11760    0.0    0.0     0.0    0.0
       isDigit                Crapto                  src/Crapto.hs:20:1-25        457           0    0.0    0.0     0.0    0.0
       lowerRot               Crapto                  src/Crapto.hs:48:1-36        449           0    0.0    0.0     0.1    0.0
        alphabetRot           Crapto                  src/Crapto.hs:39:1-89        453       53712    0.1    0.0     0.1    0.0
         indexOf              Crapto                  src/Crapto.hs:(35,1)-(36,59) 454      639124    0.0    0.0     0.0    0.0
       upperRot               Crapto                  src/Crapto.hs:45:1-36        459           0    0.0    0.0     0.0    0.0
        alphabetRot           Crapto                  src/Crapto.hs:39:1-89        460        1103    0.0    0.0     0.0    0.0
         indexOf              Crapto                  src/Crapto.hs:(35,1)-(36,59) 461       13880    0.0    0.0     0.0    0.0
     rotFib.go.n              Crapto                  src/Crapto.hs:67:9-19        450       65471    0.0    0.0     0.0    0.0
    rotChar                   Crapto                  src/Crapto.hs:(51,1)-(55,18) 435           1    0.0    0.0     0.0    0.0
     isLower                  Crapto                  src/Crapto.hs:17:1-40        436           1    0.0    0.0     0.0    0.0
     isUpper                  Crapto                  src/Crapto.hs:14:1-40        439           1    0.0    0.0     0.0    0.0
     upperRot                 Crapto                  src/Crapto.hs:45:1-36        442           0    0.0    0.0     0.0    0.0
      alphabetRot             Crapto                  src/Crapto.hs:39:1-89        443           1    0.0    0.0     0.0    0.0
       indexOf                Crapto                  src/Crapto.hs:(35,1)-(36,59) 444          12    0.0    0.0     0.0    0.0
    rotFib.(...)              Crapto                  src/Crapto.hs:62:3-15        432           1    0.0    0.0     0.0    0.0
    rotFib.h                  Crapto                  src/Crapto.hs:62:3-15        438           1    0.0    0.0     0.0    0.0
    rotFib.left               Crapto                  src/Crapto.hs:60:3-10        451           1    0.0    0.0     0.0    0.0
    rotFib.right              Crapto                  src/Crapto.hs:61:3-11        452           1    0.0    0.0     0.0    0.0
    rotFib.t                  Crapto                  src/Crapto.hs:62:3-15        433           1    0.0    0.0     0.0    0.0
  outFile                     Main                    app/Main.hs:11:5-11          428           1    0.0    0.0     0.0    0.0
```

looks like the bottleneck is `indexOf` and it's allocated `120GB` of mem, not all at once but is that OK?

```haskell
indexOf :: (Eq a) => a -> [a] -> Int
indexOf _ [] = undefined
indexOf a (x : xs) = if x == a then 0 else 1 + indexOf a xs
```

[view code in context](https://github.com/aaaaargZombies/craptography/blob/9f3e8679153def364bc223d8e9d19709af5811cd/src/Crapto.hs#L34-L36)

So rather than working out the order of the alphabet I probably need to change this into a type that has that information stored (which makes sense as we are relying on it everywhere else)

https://github.com/aaaaargZombies/craptography/blob/9f3e8679153def364bc223d8e9d19709af5811cd/src/Crapto.hs#L11-L11

https://github.com/aaaaargZombies/craptography/blob/9f3e8679153def364bc223d8e9d19709af5811cd/src/Crapto.hs#L25-L32

**Comment by aaaaargZombies** [comment]
Down to 5 seconds. I thought simply turning `fib` into a scan would solve my issue but it's looking to be more complex. IT's forcing me to explore the language a bit more though which is good. 

Using `Text` rather than `String` is helping and working directly with `Char` values rather than indexing off a reference `Alphabet` string is also helping.

profile:

```
	Mon May  6 13:44 2024 Time and Allocation Profiling Report  (Final)

	   craptography-exe +RTS -N -p -RTS -i test.md -o out.txt

	total time  =        5.00 secs   (17061 ticks @ 1000 us, 12 processors)
	total alloc = 70,757,205,800 bytes  (excludes profiling overheads)

COST CENTRE    MODULE SRC                            %time %alloc

rotFib.(...).\ Crapto src/Crapto.hs:(50,11)-(52,37)   99.9  100.0


                                                                                                       individual      inherited
COST CENTRE                   MODULE                  SRC                           no.     entries  %time %alloc   %time %alloc

MAIN                          MAIN                    <built-in>                    207           0    0.0    0.0   100.0  100.0
 CAF                          GHC.Conc.Signal         <entire-module>               388           0    0.0    0.0     0.0    0.0
 CAF                          GHC.Event.Thread        <entire-module>               373           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Encoding         <entire-module>               354           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Encoding.Iconv   <entire-module>               352           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Exception        <entire-module>               346           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.FD               <entire-module>               345           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Handle.FD        <entire-module>               343           0    0.0    0.0     0.0    0.0
 CAF                          GHC.IO.Handle.Internals <entire-module>               342           0    0.0    0.0     0.0    0.0
 CAF:defaultOpts              Main                    app/Main.hs:16:1-11           261           0    0.0    0.0     0.0    0.0
  defaultOpts                 Main                    app/Main.hs:16:1-86           420           1    0.0    0.0     0.0    0.0
 CAF:encrapt                  Crapto                  src/Crapto.hs:58:1-7          221           0    0.0    0.0     0.0    0.0
  encrapt                     Crapto                  src/Crapto.hs:58:1-20         425           1    0.0    0.0     0.0    0.0
 CAF:lowerAlphabet_rtf        Crapto                  src/Crapto.hs:13:1-13         214           0    0.0    0.0     0.0    0.0
  lowerAlphabet               Crapto                  src/Crapto.hs:13:1-25         443           1    0.0    0.0     0.0    0.0
 CAF:lowerRot_rtl             Crapto                  src/Crapto.hs:31:1-8          219           0    0.0    0.0     0.0    0.0
  lowerRot                    Crapto                  src/Crapto.hs:31:1-36         441           1    0.0    0.0     0.0    0.0
 CAF:lvl11_r2OW               Main                    <no location info>            267           0    0.0    0.0     0.0    0.0
 CAF:lvl13_r2OY               Main                    <no location info>            268           0    0.0    0.0     0.0    0.0
 CAF:lvl15_r2P0               Main                    <no location info>            269           0    0.0    0.0     0.0    0.0
 CAF:lvl1_r2OM                Main                    <no location info>            262           0    0.0    0.0     0.0    0.0
 CAF:lvl3_r2OO                Main                    <no location info>            263           0    0.0    0.0     0.0    0.0
 CAF:lvl5_r2OQ                Main                    <no location info>            264           0    0.0    0.0     0.0    0.0
 CAF:lvl7_r2OS                Main                    <no location info>            265           0    0.0    0.0     0.0    0.0
 CAF:lvl9_r2OU                Main                    <no location info>            266           0    0.0    0.0     0.0    0.0
 CAF:main1                    Main                    <no location info>            284           0    0.0    0.0     0.0    0.0
  main                        Main                    app/Main.hs:(67,1)-(81,9)     414           1    0.0    0.0     0.0    0.0
 CAF:upperAlphabet_rtg        Crapto                  src/Crapto.hs:16:1-13         215           0    0.0    0.0     0.0    0.0
  upperAlphabet               Crapto                  src/Crapto.hs:16:1-25         436           1    0.0    0.0     0.0    0.0
 CAF:upperRot_rtk             Crapto                  src/Crapto.hs:28:1-8          218           0    0.0    0.0     0.0    0.0
  upperRot                    Crapto                  src/Crapto.hs:28:1-36         434           1    0.0    0.0     0.0    0.0
 main                         Main                    app/Main.hs:(67,1)-(81,9)     415           0    0.0    0.0   100.0  100.0
  help                        Main                    app/Main.hs:11:5-8            416           1    0.0    0.0     0.0    0.0
  inFile                      Main                    app/Main.hs:8:5-10            421           1    0.0    0.0     0.0    0.0
  main.opts                   Main                    app/Main.hs:69:7-44           417           1    0.0    0.0     0.0    0.0
   parseArguments             Main                    app/Main.hs:(19,1)-(45,15)    418           5    0.0    0.0     0.0    0.0
    parseArguments.updateOpts Main                    app/Main.hs:(23,3)-(45,15)    419           4    0.0    0.0     0.0    0.0
  main.rotate                 Main                    app/Main.hs:70:7-62           423           1    0.0    0.0   100.0  100.0
   decrapt                    Main                    app/Main.hs:10:5-11           424           1    0.0    0.0     0.0    0.0
   encrapt                    Crapto                  src/Crapto.hs:58:1-20         426           0    0.0    0.0   100.0  100.0
    rotFib                    Crapto                  src/Crapto.hs:(41,1)-(55,7)   427           1    0.0    0.0   100.0  100.0
     rotFib.(...)             Crapto                  src/Crapto.hs:44:3-74         431           1    0.0    0.0     0.0    0.0
     rotFib.(...)             Crapto                  src/Crapto.hs:(47,3)-(55,7)   429           1    0.0    0.0   100.0  100.0
      rotFib.(...).\          Crapto                  src/Crapto.hs:(50,11)-(52,37) 438       65472   99.9  100.0    99.9  100.0
       rotFib.(...).\.c       Crapto                  src/Crapto.hs:51:15-32        439       65472    0.0    0.0     0.1    0.0
        rotChar               Crapto                  src/Crapto.hs:(34,1)-(38,18)  440       65472    0.0    0.0     0.0    0.0
         lowerRot             Crapto                  src/Crapto.hs:31:1-36         442           0    0.0    0.0     0.0    0.0
          alphabetRot         Crapto                  src/Crapto.hs:22:1-120        447       53712    0.0    0.0     0.0    0.0
         upperRot             Crapto                  src/Crapto.hs:28:1-36         448           0    0.0    0.0     0.0    0.0
          alphabetRot         Crapto                  src/Crapto.hs:22:1-120        449        1103    0.0    0.0     0.0    0.0
       rotFib.(...).\.n       Crapto                  src/Crapto.hs:50:15-23        444       65471    0.0    0.0     0.0    0.0
      rotChar                 Crapto                  src/Crapto.hs:(34,1)-(38,18)  433           1    0.0    0.0     0.0    0.0
       upperRot               Crapto                  src/Crapto.hs:28:1-36         435           0    0.0    0.0     0.0    0.0
        alphabetRot           Crapto                  src/Crapto.hs:22:1-120        437           1    0.0    0.0     0.0    0.0
     rotFib.h                 Crapto                  src/Crapto.hs:44:3-74         432           1    0.0    0.0     0.0    0.0
     rotFib.left              Crapto                  src/Crapto.hs:45:3-10         445           1    0.0    0.0     0.0    0.0
     rotFib.right             Crapto                  src/Crapto.hs:46:3-11         446           1    0.0    0.0     0.0    0.0
     rotFib.secret            Crapto                  src/Crapto.hs:(47,3)-(55,7)   428           1    0.0    0.0     0.0    0.0
     rotFib.t                 Crapto                  src/Crapto.hs:44:3-74         430           1    0.0    0.0     0.0    0.0
  outFile                     Main                    app/Main.hs:9:5-11            422           1    0.0    0.0     0.0    0.0
```

Current implementation

https://github.com/aaaaargZombies/craptography/blob/7520ac4af51085a3888e056257557c1fd677d515/src/Crapto.hs

**Comment by aaaaargZombies** [comment]
over the final hurdle - I did an oopsy using `snoc` which is `O(n)` and ran every operation of the fold. fixed in c203b6e 

```
total time = 0.01 secs
```
